### PR TITLE
Prefer `ArrayList` and `ArrayDeque` to `LinkedList`

### PR DESCRIPTION
**Pull request by basil**
### Problem

Per Error Prone's [`JdkObsolete`](https://errorprone.info/bugpattern/JdkObsolete) check:

> [`LinkedList`](https://docs.oracle.com/javase/8/docs/api/java/util/LinkedList.html) almost never out-performs [`ArrayList`](https://docs.oracle.com/javase/8/docs/api/java/util/ArrayList.html) or [`ArrayDeque`](https://docs.oracle.com/javase/8/docs/api/java/util/ArrayDeque.html). […] People generally choose `LinkedList` because they want fast insertion and removal. However, `LinkedList` has slow traversal, and typically you need to traverse the list to find the place to insert/remove. It turns out that the cost of traversing to a location in a `LinkedList` is approximately 4x the cost of copying an element in an `ArrayList`. Thus, `LinkedList`’s traversal cost dominates and results in poorer performance. [More info](https://stuartmarks.wordpress.com/2015/12/18/some-java-list-benchmarks/)

### Solution

When using `LinkedList` as a list, prefer `ArrayList`. When using `LinkedList` as a stack or queue/deque, prefer `ArrayDeque`.

### Implementation

Per the same article:

> Migration gotcha: `LinkedList` permits `null` elements; `ArrayDeque` rejects them. The documentation for `Deque` strongly discourages users from inserting `null`, even into implementations that permit it. So, if you are using a `LinkedList` for this purpose, you should likely stop, and you will need to stop in order to migrate to `ArrayDeque`.

The only two new usages of `ArrayDeque` I am introducing in this PR are both in `jenkins.util.VirtualFile`. In both cases, I verified that the arguments being passed to `ArrayDeque` methods (`File#getName` or `FilePath#getName`) cannot be `null`.

Note that I did _not_ touch the calls to `new LinkedList<>()` in `jenkins.widgets.HistoryPageFilter` or `hudson.widgets.BuildHistoryWidget`, because I could not easily prove that the lists never contained `null` elements and did not want to risk a regression.

### Proposed changelog entries

N/A

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by res0nance — APPROVED** [review]
Sure 🤷

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
